Audit the CHANGELOG (#1218)
spatialos/UnrealGDK: CHANGELOG.md

### Breaking Changes:
* You must [re-build](https://docs.improbable.io/unreal/alpha/content/get-started/example-project/exampleproject-setup#step-4-build-the-dependencies-and-launch-the-project) your [Example Project](https://github.com/spatialos/UnrealGDKExampleProject) if you're upgrading it to `0.6.0`.
* This is the last GDK version to support `4.20-SpatialGDK-release`. You will need to upgrade your project to use `4.22-SpatialGDK-release` if you want to continue to recieve GDK releases and support.

### New Known Issues:
- Workers will sometimes not gain authority when quickly reconnecting to an existing deployment, resulting in a failure to spawn or simulate. When using the editor if you Play - Stop - Play in quick succession you can sometimes fail to launch correctly.

### Features:

> - The GDK now uses SpatialOS `13.8.1`.

Reviewer: Dunno if this is relevant

Reviewer 2: Why? Should we say defaults to `13.8.1`?

- RPCs on `NotSpatial` types are no longer queued forever and are now dropped instead.
- Fixed issue where an Actor's Spatial position was not updated if it had an owner that was not replicated.
- BeginPlay is now only called with authority on startup actors once per deployment.
- Fixed null pointer dereference crash when trying to initiate a Spatial connection without an existing one.
- URL options are now properly sent through to the server when doing a ClientTravel.
- The correct error message is now printed when the SchemaDatabase is missing.
- `StartEditor.bat` is now generated correctly when you build a server worker outside of editor.
- Fixed an issue with logging errored blueprints after garbage collection which caused an invalid pointer crash.
- Removed the ability to configure snapshot save folder. Snapshots should always be saved to `<ProjectRoot>/spatial/snapshots`. This prevents an issue with absolute paths being checked in which can break snapshot generation.
- Introduced a new module, `SpatialGDKServices`, on which `SpatialGDK` and `SpatilGDKEditorToolbar` now depend. This resolves a previously cyclic dependency.

> - RPCs recieved on entity creation are now queued in case they cannot yet be executed. Previously they were simply dropped.

Reviewer 2: ```suggestion
- RPCs called before entity creation are now queued in case they cannot yet be executed. Previously they were simply dropped. These RPCs are also included in RPC metrics.
```

- RPCs are now guaranteed to arrive in the same order for a given actor and all of its subobjects. This matches native Unreal behavior.

> - RPCs called before entity creation are now included in RPC metrics.

Reviewer 2: ```suggestion
```

> - RPCs are now guaranteed to arrive in the same order for a given actor and all of its subobjects. This matches native Unreal behavior.

Reviewer 3: Should we specify that this is for single server?

Reviewer 2: single worker? yeah I think so

- BeginPlay is now only called with authority on startup actors once per deployment.
- Fixed null pointer dereference crash when trying to initiate a Spatial connection without an existing one.
- URL options are now properly sent through to the server when doing a ClientTravel.
- The correct error message is now printed when the SchemaDatabase is missing.
- `StartEditor.bat` is now generated correctly when you build a server worker outside of editor.
- Fixed an issue with logging errored blueprints after garbage collection which caused an invalid pointer crash.
- Removed the ability to configure snapshot save folder. Snapshots should always be saved to `<ProjectRoot>/spatial/snapshots`. This prevents an issue with absolute paths being checked in which can break snapshot generation.
- Introduced a new module, `SpatialGDKServices`, on which `SpatialGDK` and `SpatilGDKEditorToolbar` now depend. This resolves a previously cyclic dependency.
- RPCs called before entity creation are now queued in case they cannot yet be executed. Previously they were simply dropped. These RPCs are also included in RPC metrics.

> - RPCs are now guaranteed to arrive in the same order for a given actor and all of its subobjects. This matches native Unreal behavior.

Reviewer 2: ```suggestion
- RPCs are now guaranteed to arrive, on a single worker, in the same order for a given actor and all of its subobjects. This matches native Unreal behavior.
```

### Features:
- The GDK now uses SpatialOS `13.8.1`.
- Local deployment startup time has been significantly reduced.
- Local deployments now start automatically when you select `Play`. This means you no longer need to select `Start` in the GDK toolbar before you select `Play` in the Unreal toolbar.

> - If your schema has changed dureing a local deployment, the next time you select `Play` the deployment will automatically restart.

Reviewer 4: "dureing" -> "during"

Author: fixed

> - Local deployments no longer run in a seperate Command Prompt, they now run in the Unreal Editor's Output Log.

Reviewer 4: "they now run in the Unreal Editor's Output Log" -> "logs are now found in the Unreal Editor's Output Log"

Author: fixed

- Unreal Engine 4.22 is now supported. You can find the 4.22 verson of our engine fork [here](https://github.com/improbableio/UnrealEngine/tree/4.22-SpatialOSUnrealGDK-release).
- Setup.bat can now take a project path as an argument. This allows the UnrealGDK to be installed as an Engine Plugin, pass the project path as the first variable if you are running Setup.bat from UnrealEngine/Engine/Plugins.
- Removed the need for setting the `UNREAL_HOME` environment variable. The build and setup scripts will now use your project's engine association to find the Unreal Engine directory. If an association is not set they will search parent directories looking for the 'Engine' folder.

> - Added the `SpatialMetricsDisplay` class, which you can use to view UnrealWorker stats as an overlay on the client.

Reviewer 2: ```suggestion
- Added the `ASpatialMetricsDisplay` class, which you can use to view UnrealWorker stats as an overlay on the client.
```

- Added the `SpatialMetricsDisplay` class, which you can use to view UnrealWorker stats as an overlay on the client.
- Added the runtime option `bEnableHandover`, which you can use to toggle property handover when running in non-zoned deployments.
- Added the runtime option `bEnableMetricsDisplay`, which you can use to auto spawn `ASpatialMetricsDisplay`, which is used to remote debug server metrics.
- Added the runtime option `bBatchSpatialPositionUpdates`, which you can use to batch spatial position updates to the runtime.
- Started using the [schema_compiler tool](https://docs.improbable.io/reference/13.8/shared/schema/introduction#using-the-schema-compiler-directly) to generate [schema descriptors](https://docs.improbable.io/reference/13.8/shared/flexible-project-layout/build-process/schema-descriptor-build-process#schema-descriptor-introduction) rather than relying on 'spatial local launch' to do this.

> - Added the runtime option `bPackUnreliableRPCs`, which you can use to pack unreliable RPCs from the same frame in a single component update to save bandwidth.

Reviewer: @m-samiec @improbable-valentyn I would highly discourage we put this in the changelog. People shouldn't know about this I think.

Reviewer 2: Yep I agree. I'll remove.

> - Added the runtime option `bPackUnreliableRPCs`, which you can use to pack unreliable RPCs from the same frame in a single component update to save bandwidth.

Reviewer 2: ```suggestion
```

- GenerateSchemaAndSnapshots commandlet no longer runs a full schema generation for each map.
- Launching SpatialOS would fail if there was a space in the full directory path.
- Fixed an issue with schema name collisions.
- Schema generation now respects "Directories to never cook".
- The editor no longer crashes during schema generation when the database is readonly.

> - Replicating `UInterfaceProperty` no logner causes crashes.

Reviewer 2: ```suggestion
- Replicating `UInterfaceProperty` no longer causes crashes.
```
